source/transactions/tests/unified/retryable-abort-handshake.yml

```diff
@@ -0,0 +1,92 @@
+description: "retryable abortTransaction on handshake errors"
+
+schemaVersion: "1.4"
+
+runOnRequirements:
+  - minServerVersion: "4.2"
+    topologies: [replicaset, sharded, load-balanced]
+    serverless: "forbid"
+    auth: true
+
+createEntities:
+  - client:
+      id: &client0 client0
+      useMultipleMongoses: false
+      observeEvents: [commandStartedEvent, connectionCheckOutStartedEvent]
+      uriOptions: { retryReads: true, retryWrites: true, maxPoolSize: 5 }
+  - database:
+      id: &database0 database0
+      client: *client0
+      databaseName: &databaseName retryable-handshake-tests
+  - collection:
+      id: &collection0 collection0
+      database: *database0
+      collectionName: &collectionName coll
+  - session:
+      id: &session0 session0
+      client: *client0
+
+initialData:
+  - collectionName: *collectionName
+    databaseName: *databaseName
+    documents:
+      - { _id: 1, x: 11 }
+
+tests:
+  - description: "AbortTransaction succeeds after handshake network error"
+    operations:
+
+      - name: startTransaction
+        object: *session0
+
+      - name: insertOne
+        object: *collection0
+        arguments:
+          document: { _id: 2, x: 22 }
+
+      - name: failPoint # fail the next connection establishment
```

Comment:
Why does this refer to "next connection". Is the subsequent `abortTransaction` going to check a new connection out of the pool?

I read through [Checking Out a Connection](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#id32) in the CMAP spec but it's still not clear to me how the pool orders its connections for re-use.

Reply:
Good Q here, my phrasing is strange, I need to give these transaction tests a bit more thought as opposed to the normal retryable ones. The other tests rely on connection pools being started with zero connections, so we can bank on the first operation being the one to perform the handshake. Which is why we needed to have the pool drop the connection created by the failPoint so we get back down to zero connections. 

With transaction tests, I need to think about how I can engineer the scenario where the pool is emptied before we run an abort/commit so that those operations would run a handshake. In reality I suppose this could happen due to some networking errors that reset the pool(s) before the abort/commit is run. 🤔

Reply:
I've updated the transaction tests now, Shane cleverly suggested using multiple sessions to avoid killing the transaction before we're able to run abort or commit. Now the dropCollections trick should empty the pool right before those two final operations.